File: module2/class1/exemplo4.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 4 - HOTSPOT EM ALGORITMO DE ORDENAÇÃO (C++)
================================================================================

OBJETIVO:
- Implementar algoritmo de ordenação simples (Bubble Sort) em grande vetor
- Usar CPU Usage Tool para identificar impacto do algoritmo e pontos quentes
- Substituir por QuickSort ou algoritmo mais eficiente
- Comparar perfis e medir ganho de performance

PROBLEMA:
- Bubble Sort possui complexidade O(n²) com loops aninhados
- CPU Usage Tool mostrará hotspot nos loops aninhados da função bubbleSort
- Para 25.000 elementos: ~625 milhões de operações teóricas

SOLUÇÃO:
- Implementar QuickSort com complexidade O(n log n)
- Resultado: redução de 625M para ~367K operações (1000x+ melhoria)

================================================================================
*/

#include <iostream>
#include <chrono>
#include <random>
#include <algorithm>
using namespace std;

void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n-1; i++) {
        for (int j = 0; j < n-i-1; j++) { // CPU HOTSPOT: O(n²) nested loops - Replace with QuickSort O(n log n) for better performance
            if (arr[j] > arr[j+1]) {
                swap(arr[j], arr[j+1]);
            }
        }
        
        if (i % 1000 == 0) {
            cout << "Bubble Sort Progress: " << i << "/" << (n-1) << " passes completed" << endl;
        }
    }
}

int main() {
    const int ARRAY_SIZE = 25000;
    cout << "Starting CPU hotspot demonstration with Bubble Sort..." << endl;
    cout << "Array size: " << ARRAY_SIZE << " elements" << endl;
    cout << "Monitor CPU usage - Bubble Sort will show O(n²) complexity hotspot" << endl;
    
    int* arr = new int[ARRAY_SIZE];
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(1, 100000);
    
    for (int i = 0; i < ARRAY_SIZE; i++) {
        arr[i] = dis(gen);
    }
    
    cout << "Array generated with random values" << endl;
    cout << "Starting Bubble Sort..." << endl;
    
    auto start = chrono::high_resolution_clock::now();
    
    bubbleSort(arr, ARRAY_SIZE);
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << endl;
    cout << "=== SORTING COMPLETED ===" << endl;
    cout << "Bubble Sort execution time: " << duration.count() << " ms" << endl;
    cout << "Algorithm complexity: O(n²) - " << (ARRAY_SIZE * ARRAY_SIZE / 1000000.0) << "M theoretical operations" << endl;
    
    bool isSorted = is_sorted(arr, arr + ARRAY_SIZE);
    cout << "Array is " << (isSorted ? "correctly sorted" : "NOT sorted") << endl;
    
    delete[] arr;
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR QUICKSORT OTIMIZADO)
================================================================================

#include <iostream>
#include <chrono>
#include <random>
#include <algorithm>
using namespace std;

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return (i + 1);
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);  // CORREÇÃO: O(n log n) complexity instead of O(n²)
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    const int ARRAY_SIZE = 25000;
    cout << "Starting optimized sorting demonstration with QuickSort..." << endl;
    cout << "Array size: " << ARRAY_SIZE << " elements" << endl;
    cout << "Monitor CPU usage - QuickSort will show O(n log n) complexity" << endl;
    
    int* arr = new int[ARRAY_SIZE];
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(1, 100000);
    
    for (int i = 0; i < ARRAY_SIZE; i++) {
        arr[i] = dis(gen);
    }
    
    cout << "Array generated with random values" << endl;
    cout << "Starting QuickSort..." << endl;
    
    auto start = chrono::high_resolution_clock::now();
    
    quickSort(arr, 0, ARRAY_SIZE - 1);
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << endl;
    cout << "=== SORTING COMPLETED ===" << endl;
    cout << "QuickSort execution time: " << duration.count() << " ms" << endl;
    cout << "Algorithm complexity: O(n log n) - " << (ARRAY_SIZE * log2(ARRAY_SIZE) / 1000.0) << "K theoretical operations" << endl;
    
    bool isSorted = is_sorted(arr, arr + ARRAY_SIZE);
    cout << "Array is " << (isSorted ? "correctly sorted" : "NOT sorted") << endl;
    
    delete[] arr;
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo4.cs
/*
================================================================================
ATIVIDADE PRÁTICA 4 - HOTSPOT EM ALGORITMO DE ORDENAÇÃO (C#)
================================================================================

OBJETIVO:
- Implementar algoritmo de ordenação simples (Bubble Sort) em grande array
- Usar CPU Usage Tool para identificar impacto do algoritmo e pontos quentes
- Substituir por QuickSort ou algoritmo mais eficiente
- Comparar perfis e medir ganho de performance

PROBLEMA:
- Bubble Sort possui complexidade O(n²) com loops aninhados
- CPU Usage Tool mostrará hotspot nos loops aninhados do método BubbleSort
- Para 25.000 elementos: ~625 milhões de operações teóricas

SOLUÇÃO:
- Implementar QuickSort com complexidade O(n log n)
- Resultado: redução de 625M para ~367K operações (1000x+ melhoria)

================================================================================
*/

using System;
using System.Diagnostics;

class Program {
    static void BubbleSort(int[] arr) {
        int n = arr.Length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) { // CPU HOTSPOT: O(n²) nested loops - Replace with QuickSort O(n log n) for better performance
                if (arr[j] > arr[j + 1]) {
                    (arr[j], arr[j + 1]) = (arr[j + 1], arr[j]);
                }
            }
            
            if (i % 1000 == 0) {
                Console.WriteLine($"Bubble Sort Progress: {i}/{n-1} passes completed");
            }
        }
    }
    
    static void Main() {
        const int ARRAY_SIZE = 25000;
        Console.WriteLine("Starting CPU hotspot demonstration with Bubble Sort...");
        Console.WriteLine($"Array size: {ARRAY_SIZE} elements");
        Console.WriteLine("Monitor CPU usage - Bubble Sort will show O(n²) complexity hotspot");
        
        int[] arr = new int[ARRAY_SIZE];
        Random rnd = new Random();
        
        for (int i = 0; i < ARRAY_SIZE; i++) {
            arr[i] = rnd.Next(1, 100001);
        }
        
        Console.WriteLine("Array generated with random values");
        Console.WriteLine("Starting Bubble Sort...");
        
        var sw = Stopwatch.StartNew();
        
        BubbleSort(arr);
        
        sw.Stop();
        
        Console.WriteLine();
        Console.WriteLine("=== SORTING COMPLETED ===");
        Console.WriteLine($"Bubble Sort execution time: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Algorithm complexity: O(n²) - {(ARRAY_SIZE * ARRAY_SIZE / 1000000.0):F1}M theoretical operations");
        
        bool isSorted = true;
        for (int i = 0; i < ARRAY_SIZE - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                isSorted = false;
                break;
            }
        }
        Console.WriteLine($"Array is {(isSorted ? "correctly sorted" : "NOT sorted")}");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR QUICKSORT OTIMIZADO)
================================================================================

using System;
using System.Diagnostics;

class Program {
    static int Partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);
        
        for (int j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;
                (arr[i], arr[j]) = (arr[j], arr[i]);
            }
        }
        (arr[i + 1], arr[high]) = (arr[high], arr[i + 1]);
        return (i + 1);
    }
    
    static void QuickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = Partition(arr, low, high);
            QuickSort(arr, low, pi - 1);  // CORREÇÃO: O(n log n) complexity instead of O(n²)
            QuickSort(arr, pi + 1, high);
        }
    }
    
    static void Main() {
        const int ARRAY_SIZE = 25000;
        Console.WriteLine("Starting optimized sorting demonstration with QuickSort...");
        Console.WriteLine($"Array size: {ARRAY_SIZE} elements");
        Console.WriteLine("Monitor CPU usage - QuickSort will show O(n log n) complexity");
        
        int[] arr = new int[ARRAY_SIZE];
        Random rnd = new Random();
        
        for (int i = 0; i < ARRAY_SIZE; i++) {
            arr[i] = rnd.Next(1, 100001);
        }
        
        Console.WriteLine("Array generated with random values");
        Console.WriteLine("Starting QuickSort...");
        
        var sw = Stopwatch.StartNew();
        
        QuickSort(arr, 0, ARRAY_SIZE - 1);
        
        sw.Stop();
        
        Console.WriteLine();
        Console.WriteLine("=== SORTING COMPLETED ===");
        Console.WriteLine($"QuickSort execution time: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Algorithm complexity: O(n log n) - {(ARRAY_SIZE * Math.Log2(ARRAY_SIZE) / 1000.0):F1}K theoretical operations");
        
        bool isSorted = true;
        for (int i = 0; i < ARRAY_SIZE - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                isSorted = false;
                break;
            }
        }
        Console.WriteLine($"Array is {(isSorted ? "correctly sorted" : "NOT sorted")}");
    }
}

================================================================================
*/